flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/TriggerManager.java

```diff
@@ -0,0 +1,295 @@
+        LOG.debug("Nothing to execute at {} for collected: {}", current, changes);
+      }
+
+      startsFrom = 0;
+      return;
+    }
+
+    if (lock.tryLock()) {
+      TableChange change = changes.get(taskToStart);
+      SerializableTable table =
+          (SerializableTable) SerializableTable.copyOf(tableLoader.loadTable());
+      out.collect(Trigger.create(current, table, taskToStart));
+      LOG.debug("Fired event with time: {}, collected: {} for {}", current, change, table.name());
+      triggerCounters.get(taskToStart).inc();
+      changes.set(taskToStart, TableChange.empty());
+      accumulatedChanges.update(changes);
+      schedule(timerService, current + minFireDelayMs);
+      startsFrom = taskToStart + 1;
+    } else {
+      // The lock is already held by someone
+      LOG.info("Delaying task on failed lock check: {}", current);
+
+      startsFrom = taskToStart;
+      concurrentRunTriggeredCounter.inc();
+      schedule(timerService, current + lockCheckDelayMs);
+    }
+
+    timerService.registerProcessingTimeTimer(nextEvaluationTime.value());
+  }
+
+  private void schedule(TimerService timerService, long time) throws IOException {
+    nextEvaluationTime.update(time);
+    timerService.registerProcessingTimeTimer(time);
+  }
+
+  private static Integer nextTrigger(
+      List<TriggerEvaluator> evaluators,
+      List<TableChange> changes,
+      List<Long> lastTriggerTimes,
+      long currentTime,
+      int startPos) {
+    int normalizedStartingPos = startPos % evaluators.size();
+    int current = normalizedStartingPos;
+    do {
+      if (evaluators
+          .get(current)
+          .check(changes.get(current), lastTriggerTimes.get(current), currentTime)) {
+        return current;
+      }
+
+      current = (current + 1) % evaluators.size();
+    } while (current != normalizedStartingPos);
+
+    return null;
+  }
+
+  private void init(Collector<Trigger> out, TimerService timerService, long current)
+      throws Exception {
+    if (!inited) {
+      // Initialize with empty changes and current timestamp
+      if (!accumulatedChanges.get().iterator().hasNext()) {
+        List<TableChange> changes = Lists.newArrayListWithCapacity(evaluators.size());
+        List<Long> triggerTimes = Lists.newArrayListWithCapacity(evaluators.size());
+        for (int i = 0; i < evaluators.size(); ++i) {
+          changes.add(TableChange.empty());
+          triggerTimes.add(current);
+        }
+
+        accumulatedChanges.update(changes);
+        lastTriggerTimes.update(triggerTimes);
+      }
+
+      if (isCleanUp) {
+        // When the job state is restored, there could be ongoing tasks.
+        // To prevent collision with the new triggers the following is done:
+        //  - add a cleanup lock
+        //  - fire a clean-up trigger
+        // This ensures that the tasks of the previous trigger are executed, and the lock is removed
+        // in the end.
+        recoveryLock.tryLock();
```

Comment:
where is recoveryLock unlocked?

Reply:
In the last operator of the Table Maintenance Stream, where we join the results of all of the Maintenance Tasks, and remove the lock when the Watermark has arrived from all of them. This makes sure that any recovered and still running task is finished at this time.

We essentially use the Watermark as a marker that every record is processed in all of the Maintenance Tasks.

Reply:
Yeah
- If we succeed, then we were able to create the lock and this is fine
- If we fail, then we already have a lock on the table (maybe from a previous failed recovery), and this is fine as well